helix-loader/src/grammar.rs

```diff
@@ -0,0 +1,388 @@
+use anyhow::{anyhow, Context, Result};
+use libloading::{Library, Symbol};
+use serde::{Deserialize, Serialize};
+use std::fs;
+use std::time::SystemTime;
+use std::{
+    collections::HashSet,
+    path::{Path, PathBuf},
+    process::Command,
+    sync::mpsc::channel,
+};
+use tree_sitter::Language;
+
+#[cfg(unix)]
+const DYLIB_EXTENSION: &str = "so";
+
+#[cfg(windows)]
+const DYLIB_EXTENSION: &str = "dll";
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Configuration {
+    #[serde(rename = "use-grammars")]
+    pub grammar_selection: Option<GrammarSelection>,
+    pub grammar: Vec<GrammarConfiguration>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase", untagged)]
+pub enum GrammarSelection {
+    Only(HashSet<String>),
+    Except(HashSet<String>),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct GrammarConfiguration {
+    #[serde(rename = "name")]
+    pub grammar_id: String,
+    pub source: GrammarSource,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase", untagged)]
+pub enum GrammarSource {
+    Local {
+        path: String,
+    },
+    Git {
+        #[serde(rename = "git")]
+        remote: String,
+        #[serde(rename = "rev")]
+        revision: String,
+        subpath: Option<String>,
+    },
```

Comment:
Ah, even though I switched `Git` and `Local` position it still did not work, I guess if we need it then we need to do a custom deserializer, seemed like cargo also have a custom one. https://github.com/rust-lang/cargo/blob/ef3294d93d82a7fe05233d7de3aa7dddf7e780fe/src/cargo/core/dependency.rs#L55-L76

Reply:
ah that makes sense:

> Serde will try to match the data against each variant in order and the first one that deserializes successfully is the one returned.

should we try out using a custom deserializer? I'm ok with subpath and it's not very common, only 4 gramars use it currently

Reply:
Yeah I'm fine with `subpath`, it's more clear that it's a git dependency with a subfolder being used. If both options used a `path`, this could look like a git dependency with a local path override

Reply:
I am fine with `subpath` too but `path` allows developers to take knowledge from existing cargo so it's straightforward and developer don't need to learn anything new if coming from rust.

Reply:
Cargo doesn't actually support `{ git = "..", path = ".." }` though, I think you're thinking of `{ git = "..", package = ".." }`